components/board/genericlinux/gpio.go

```diff
@@ -231,12 +224,24 @@ func (pin *gpioPin) startSoftwarePWM() error {
 	// If we get here, we need a software loop to drive the PWM signal, either because this pin
 	// doesn't have hardware support or because we want to drive it at such a low frequency that
 	// the hardware chip can't do it.
-	if pin.softwarePwm != nil {
+	if pin.enableSoftwarePWM {
 		// We already have a software PWM loop running. It will pick up the changes on its own.
 		return nil
 	}
 
-	pin.softwarePwm = rdkutils.NewStoppableWorkers(pin.softwarePwmLoop)
+	pin.enableSoftwarePWM = true
+	// Sneaky trick alert! We use startSoftwarePWMChan to tell the background worker to start up.
+	// However, we have acquired the mutex, and if the background worker was already started
+	// (because it had been started, another goroutine told it to stop, and we are running so
+	// soon afterwards that it hasn't yet noticed it should stop), it's possible the background
+	// worker is still running and is about to acquire the mutex. If we tried sending a message on
+	// the channel in that situation, we would deadlock (we would wait until something can read
+	// from the channel, and the only goroutine that can read from it would wait until the mutex is
+	// unlocked). So, instead, we send the message by closing the old channel and creating a new
+	// one: if the background worker was waiting on the channel, it will wake up when it closes,
+	// and if it was not waiting on the channel, we don't block waiting for it to start waiting.
+	close(pin.startSoftwarePWMChan)
+	pin.startSoftwarePWMChan = make(chan any)
```

Comment:
ugh. [Assigning a value is not atomic](https://viam.atlassian.net/browse/RSDK-8345) in Go. We've locked the mutex for this line, but we need to also lock it when we read from the channel. but we can't lock the mutex for the select statement (which might take multiple hours to complete). So, the background goroutine needs to lock the mutex, grab a pointer to the channel, unlock the mutex, and then select on the pointer.